src/redux/gas.ts

```diff
@@ -0,0 +1,591 @@
+              }
+            }
+
+            if (!isEmpty(existingGasFees[CUSTOM])) {
+              // Preserve custom values while updating prices
+              gasFeeParamsBySpeed[CUSTOM] = {
+                ...existingGasFees[CUSTOM],
+                baseFeePerGas: baseFee,
+              };
+            } else if (isEmpty(gasFeeParamsBySpeed[CUSTOM])) {
+              // set CUSTOM to NORMAL if not defined
+              // gasFeeParamsBySpeed[CUSTOM] = {} as GasFeeParamsBySpeed;
+            }
+            dispatch({
+              payload: {
+                confirmationTimeByPriorityFee,
+                currentBlockParams: { baseFeePerGas: currentBaseFee, trend },
+                gasFeeParamsBySpeed: gasFeeParamsBySpeed,
+              },
+              type: GAS_FEES_SUCCESS,
+            });
+          } catch (e) {
+            captureException(new Error('Etherscan gas estimates failed'));
+            logger.sentry('Etherscan gas estimates error:', e);
+            logger.sentry('falling back to eth gas station');
+          }
+        }
+        fetchResolve(true);
+      } catch (error) {
+        const fallbackGasPrices = getFallbackGasPrices();
+        captureException(new Error('all gas estimates failed'));
+        logger.sentry('gas estimates error', error);
+        dispatch({
+          payload: fallbackGasPrices,
+          type: GAS_PRICES_FAILURE,
+        });
+        fetchReject(error);
+      }
+    });
+
+  const watchGasPrices = async (network: Network) => {
+    try {
+      await getGasPrices(network);
+      // eslint-disable-next-line no-empty
+    } catch (e) {
+    } finally {
+      gasPricesHandle = setTimeout(() => {
+        watchGasPrices(network);
+      }, 15000); // 15 secs
+    }
+  };
+
+  watchGasPrices(network);
+};
+
+export const gasUpdateGasFeeOption = (
+  newGasPriceOption: string,
+  assetsOverride?: any[]
+) => (dispatch: AppDispatch, getState: AppGetState) => {
+  const {
+    gasFeeParamsBySpeed,
+    gasFeesBySpeed,
+    txNetwork,
+    selectedGasFee: oldSelectedFee,
+  } = getState().gas;
+  if (oldSelectedFee.option === newGasPriceOption) return;
+
+  const { assets } = getState().data;
+
+  const gasPriceOption = newGasPriceOption || NORMAL;
+  if (isEmpty(gasFeeParamsBySpeed)) return;
+  const selectedGasFee = getSelectedGasFee(
+    assetsOverride || assets,
+    gasFeeParamsBySpeed,
+    gasFeesBySpeed,
+    gasPriceOption,
+    txNetwork
+  );
+  dispatch({
+    payload: selectedGasFee,
+    type: GAS_UPDATE_GAS_PRICE_OPTION,
+  });
+  analytics.track('Updated Gas Price', { gasPriceOption: gasPriceOption });
+};
+
+export const gasUpdateDefaultGasLimit = (
+  defaultGasLimit = ethUnits.basic_tx
+) => (dispatch: AppDispatch) => {
+  dispatch({
+    payload: defaultGasLimit,
+    type: GAS_UPDATE_DEFAULT_GAS_LIMIT,
+  });
+  dispatch(gasUpdateTxFee(defaultGasLimit));
+};
+
+export const gasUpdateTxFee = (
+  gasLimit: number,
```

Comment:
type is `gasLimit?: number` as gasLimit can be undefined and it will fallback to `defaultGasLimit`